legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/test/TestModelBuilder.java

```diff
@@ -92,17 +92,30 @@ public TestModelBuilder(DependenciesApi dependenciesApi, EntityApi entityApi, Pr
     {
         try
         {
+            Set<String> latestUpstreamDependenciesIds = SetAdapter.adapt(latestUpstreamDependencies).collect(dependency -> dependency.getDepotProjectId().toString());
+            if (latestUpstreamDependenciesIds.contains(downstreamProjectId))
+            {
+                throw new IllegalArgumentException("Project " + downstreamProjectId + " was specified as downstream but in fact it is a direct dependency for upstream project " + upstreamProjectId.toString());
+            }
```

Comment:
There's no point in building a set of all the dependency ids just to check if one is in there. Just check on the original set. For that matter, it's probably better for downstreamProjectId to be parsed to a DepotProjectId at this point.